### ISSUE TITLE
ESLint doesn't recognize doc comments as a use of a type

### ISSUE THREAD

**Issue by v-kydela** [eslint/eslint#12869]
**Tell us about your environment**

* **ESLint Version:** v6.8.0
* **Node Version:** v8.11.3
* **npm Version:** v5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    "extends": "standard",
    "rules": {    
        "semi": [2, "always"],
        "indent": [2, 4],
        "no-return-await": 0,
        "space-before-function-paren": [2, {
            "named": "never",
            "anonymous": "never",
            "asyncArrow": "always"
        }],
        "template-curly-spacing": [2, "always"]
    }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const { TurnContext } = require('botbuilder');
/**
 *
 * @param {TurnContext} turnContext on turn context object.
 */
async run(turnContext) {}
```

`eslint test.js`

**What did you expect to happen?**
ESLint should recognize that the types I'm importing are used in doc comments

**What actually happened? Please include the actual, raw output from ESLint.**
```
6:66  error  'TurnContext' is assigned a value but never used        no-unused-vars
```

**Are you willing to submit a pull request to fix this bug?**
Maybe

**Comment by mdjermanovic** [comment]
Hi @v-kydela, thanks for the bug report!

However, this is out of scope for this rule as there is no such thing as importing types in JS.

Is there an alternative way to refer to a type, instead of declaring variables and using `require`?

I think this should work for VSCode IntelliSense:

```js
/**
 * @typedef {import('botbuilder').TurnContext} TurnContext
 */
```

**Comment by v-kydela** [comment]
@mdjermanovic - Your import trick does work so thank you for showing me that. However, that unfortunately doesn't adhere to the conventions used by my organization.

I'm sorry if I misspoke by using the word "type." JavaScript does have classes, though, and the class is being required so that it can be used by IntelliSense. Is there any chance the scope of the `no-unused-vars` rule could be changed to accommodate this?

**Comment by mdjermanovic** [comment]
> I'm sorry if I misspoke by using the word "type." JavaScript does have classes, though, and the class is being required so that it can be used by IntelliSense. Is there any chance the scope of the `no-unused-vars` rule could be changed to accommodate this?

Thanks for the clarification!

I'm not sure that we'll be able to support this as it would require parsing JSDoc comments (and supporting JSDoc in general), which has been [deprecated from the core](https://eslint.org/blog/2018/11/jsdoc-end-of-life) for maintenance reasons.

Marked this issue as an enhancement request, since other team members may have a different opinion.

**Comment by kaicataldo** [comment]
@v-kydela I don't think we'll be able to support this in core. There are two existing plugins that might be a good place to propose something like this:

- https://www.npmjs.com/package/eslint-plugin-jsdoc
- https://www.npmjs.com/package/@typescript-eslint/eslint-plugin

**Comment by v-kydela** [comment]
@kaicataldo - Is the JSDoc plugin meant to be used on top of ESLint or instead of? It seems to me that any additional plugins that run alongside ESLint could not fix this problem, because no matter how some other plugin lints JSDoc comments ESLint would still not see the use of the class in the comment.

**Comment by kaicataldo** [comment]
> Is the JSDoc plugin meant to be used on top of ESLint or instead of?

Assuming you mean using a JSDoc plugin rule in addition to `no-unused-vars` in ESLint core, it should be possible to use it in addition. This should be doable similarly to how `eslint-plugin-react` marks `import React from "react"` as used in the `jsx-uses-react` rule.

Docs are [https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-uses-react.md] and the rule source code can be found [here](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-uses-react.js).
